src/main/java/org/broadinstitute/hellbender/utils/hdf5/HDF5PoNCreator.java

```diff
@@ -0,0 +1,712 @@
+        final int targetCount = counts.getRowDimension();
+        final int columnCount = counts.getColumnDimension();
+
+        // Declared once outside loop to reuse across rows:
+        final int[] zeroIndexes = new int[columnCount];
+        final double[] nonZeroValues = new double[columnCount];
+        final Median medianCalculator = new Median();
+        int totalCounts = 0;
+        int totalZeroCounts = 0;
+
+        // for each row aka target:
+        for (int i = 0; i < targetCount; i++) {
+            // We classify columns as having a zero or not.
+            // we accumulate nonZero values in nonZeroValues to calculate the median
+            // where as we accumulate zero indexes that need to be updated later.
+            final double[] rowCounts = counts.getRow(i);
+            int zeroCount = 0;
+            int nonZeroCount = 0;
+            for (int j = 0; j <  rowCounts.length; j++) {
+                if (rowCounts[j] != 0.0) {
+                    nonZeroValues[nonZeroCount++] = rowCounts[j];
+                } else {
+                    zeroIndexes[zeroCount++] = j;
+                }
+            }
+            totalCounts += rowCounts.length;
+            if (zeroCount > 0) { // We only have to do anything if if there is any zeros in the row.
+                final double nonZeroMedian = medianCalculator.evaluate(nonZeroValues, 0, nonZeroCount);
+                for (int j = 0; j < zeroCount; j++) {
+                    counts.setEntry(i, zeroIndexes[j], nonZeroMedian);
+                }
+                totalZeroCounts += zeroCount;
+            }
+        }
+        if (totalZeroCounts > 0) {
+            logger.info(String.format("Some 0.0 counts, %d of %d (%.2f%%), where imputed to their enclosing target non-zero median", totalZeroCounts, totalZeroCounts, 100.0 * (totalZeroCounts / totalCounts)));
+        } else {
+            logger.info("No count is 0.0 thus no count needed to be imputed.");
+        }
+    }
+    /**
+     * Truncates the extreme count values in the input read-count collection.
+     * Values are forced to be bound by the percentile indicated with the input {@code percentile} which must be
+     * in the range [0 .. 50.0]. Values under that percentile and the complementary (1 - percentile) are set to the
+     * corresponding threshold value.
+     *
+     * <p>The imputation is done in-place, thus the input matrix is well be modified as a result of this call.</p>
+     *
+     * @param readCounts the input and output read-count matrix.
+     */
+    @VisibleForTesting
+    static void truncateExtremeCounts(final ReadCountCollection readCounts, final double percentile, final Logger logger) {
+
+        final RealMatrix counts = readCounts.counts();
+        final int targetCount = counts.getRowDimension();
+        final int columnCount = counts.getColumnDimension();
+        final double[] values = new double[targetCount * columnCount];
```

Comment:
How about something like `double[] values = IntStream.range(0, targetCount).flatMap(t -> counts.getRow(t))`?  (I'm sure I'm forgetting `boxed()` etc. but it still ought to be a one-liner).